src/rpc/blockchain.cpp

```diff
@@ -444,12 +444,18 @@ UniValue getrawmempool(const JSONRPCRequest& request)
             + HelpExampleCli("getrawmempool", "true")
             + HelpExampleRpc("getrawmempool", "true")
         );
+    }
 
     bool fVerbose = false;
-    if (!request.params[0].isNull())
+    if (!request.params[0].isNull()) {
         fVerbose = request.params[0].get_bool();
+    }
+
+    const UniValue txpool_json{mempoolToJSON(fVerbose)};
 
-    return mempoolToJSON(fVerbose);
+    // Wait for ATMP calling thread to release the write lock:
+    LOCK(cs_main);
```

Comment:
This needs to go before mempoolToJONS, no? ie just do a { LOCK(cs_main); }.

Reply:
Acquiring the lock and immediately giving it back to the ATMP calling thread will not solve any races. (At least on my machine I can still see them.)

I guess, I could hold it for the whole duration of getrawmempool, if that is what you like.

Reply:
Ah, oops, yes, sorry, indeed, wrong direction. I guess this is fine for now, but we really need to kill the mempool.cs/cs_main garbage.

Reply:
I don't understand why does it matter waiting for `cs_main` when the result is already determined.

Reply:
We want to wait for https://github.com/bitcoin/bitcoin/blob/8470e64724cb4a668079e7516dc61ec230346618/src/validation.cpp#L974
which happens under `cs_main`, not `pool.cs`.